### DISCUSSION TITLE
[@types/ramda] Path type doesn't support symbols

### DISCUSSION THREAD

**Discussion by Jazzmanpw** [DefinitelyTyped/DefinitelyTyped#64273] [Issues with a @types package]
There are some ramda functions that work on paths ([assocPath](https://ramdajs.com/docs/#assocPath), [hasPath](https://ramdajs.com/docs/#hasPath) and others). At the runtime, they work with strings, numbers as well as symbols, but `type Path` used as the argument is `ReadonlyArray<string | number>`, but should be `ReadonlyArray<string | number | symbol>`.

It seems like a trivial fix, but I have no idea how to test the changes automatically, and I don't have time to dig into this, so I wonder if it could be a good first issue, or someone experienced could fix it quicker than me.

P.S. Regarding tests. There's construction `R.assocPath(R.__, 42, testObj)(testPath); // => {x: [{y: 42, z: 3}, {y: 4, z: 5}]}`, with a `=>` comment, but it seems that it jsut describes the result rather than checks the returned type. And maybe it should be fixed as well.

P.P.S. For now, I just changed the type in-place for the exptected currying. Not the best option, but it works.

```ts
const mobileWidgetsSymbol = Symbol('core-mobile-widgets');
const assocPathWithSymbols = assocPath as <T, U>(
  path: readonly (number | string | symbol)[],
) => _.F.Curry<(a: T, b: U) => U>;
const extendWithMobileWidgets = assocPathWithSymbols<
  MobileWidgetProps[],
  PartialApp
>(['extensions', mobileWidgetsSymbol]);
```

[a comment by an automated account is not shown]

**Comment by Nicholaiii** [comment]
> There are some ramda functions that work on paths ([assocPath](https://ramdajs.com/docs/#assocPath), [hasPath](https://ramdajs.com/docs/#hasPath) and others). At the runtime, they work with strings, numbers as well as symbols, but `type Path` used as the argument is `ReadonlyArray<string | number>`, but should be `ReadonlyArray<string | number | symbol>`.
> 
> It seems like a trivial fix, but I have no idea how to test the changes automatically, and I don't have time to dig into this, so I wonder if it could be a good first issue, or someone experienced could fix it quicker than me.
> 
> P.S. Regarding tests. There's construction `R.assocPath(R.__, 42, testObj)(testPath); // => {x: [{y: 42, z: 3}, {y: 4, z: 5}]}`, with a `=>` comment, but it seems that it jsut describes the result rather than checks the returned type. And maybe it should be fixed as well.
> 
> P.P.S. For now, I just changed the type in-place for the exptected currying. Not the best option, but it works.
> 
> ```ts
> const mobileWidgetsSymbol = Symbol('core-mobile-widgets');
> const assocPathWithSymbols = assocPath as <T, U>(
>   path: readonly (number | string | symbol)[],
> ) => _.F.Curry<(a: T, b: U) => U>;
> const extendWithMobileWidgets = assocPathWithSymbols<
>   MobileWidgetProps[],
>   PartialApp
> >(['extensions', mobileWidgetsSymbol]);
> ```

Instead of `ReadonlyArray<string | number | symbol>` it should be `ReadonlyArray<keyof any>` as that is defined as all valid key types, which may or may not change (but rn is `string | number | symbol`).
The type comes from the tools.d.ts-provided `Path` that looks like this:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/ramda/tools.d.ts#L451

**Comment by Jazzmanpw** [comment]
I don't like `any` because TS ESlint always complains about it. It seems like `keyof never` gives the same result, but for ramda types `any` can be ok
